From 59611329805ba60ffe7e4d33580d1f89d5ddac71 Mon Sep 17 00:00:00 2001
From: "Alexander F. Lent" <lx@xanderlent.com>
Date: Sat, 12 Apr 2025 23:11:37 -0400
Subject: [PATCH] Add SKIP_KMOD/nokmod build option (fixes #342)

This allows the project to be built without building the kernel module.

Signed-off-by: Alexander F. Lent <lx@xanderlent.com>
---
 CMakeLists.txt            |  1 +
 build/build.sh            |  9 +++++++++
 src/driver/CMakeLists.txt | 13 +++++++++++--
 3 files changed, 21 insertions(+), 2 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 1301800..cae0a10 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -4,6 +4,7 @@
 cmake_minimum_required(VERSION 3.19.0)
 project("amd-xdna")
 set(PROJECT_DESCRIPTION "AMD XDNA driver and SHIM for Linux")
+option(SKIP_KMOD "Skip building the kernel module" OFF)
 
 if (DEFINED ENV{XRT_PLUGIN_VERSION_PATCH})
   SET(XRT_PLUGIN_VERSION_PATCH $ENV{XRT_PLUGIN_VERSION_PATCH})
diff --git a/build/build.sh b/build/build.sh
index 5387d3f..f9c86fb 100755
--- a/build/build.sh
+++ b/build/build.sh
@@ -23,6 +23,7 @@ Options:
 	-verbose								Enable verbose build
 	-hello_umq              Hello UMQ Memory Test
 	-dir	      						Download directory if apply
+  -nokmod									Don't build or install the kernel module
 USAGE_END
 }
 
@@ -123,6 +124,7 @@ package=0
 example=0
 nocmake=0
 verbose=
+skip_kmod=
 njobs=`grep -c ^processor /proc/cpuinfo`
 download_dir=
 xrt_install_prefix="/opt/xilinx"
@@ -172,6 +174,9 @@ while [ $# -gt 0 ]; do
 		-verbose)
 			verbose=VERBOSE=1
 			;;
+		-nokmod)
+		  skip_kmod="ON"
+			;;
 		-dir)
 			download_dir=$2
 			shift
@@ -216,6 +221,10 @@ fi
 
 cmake_extra_flags+=" -DCMAKE_INSTALL_PREFIX=$xrt_install_prefix -DXRT_INSTALL_PREFIX=$xrt_install_prefix"
 
+if [[ ! -z "$skip_kmod" ]]; then
+	cmake_extra_flags+=" -DSKIP_KMOD=$skip_kmod"
+fi
+
 if [[ ! -z "$download_dir" ]]; then
 	echo "Specified download directory is $download_dir"
 	DOWNLOAD_BINS_DIR=$download_dir
diff --git a/src/driver/CMakeLists.txt b/src/driver/CMakeLists.txt
index 9d174e6..f287e7e 100644
--- a/src/driver/CMakeLists.txt
+++ b/src/driver/CMakeLists.txt
@@ -18,6 +18,8 @@ endif(DEFINED ENV{XDNA_DRV_BLD_FLAGS})
 
 message("-- Extra driver source dir: ${XDNA_EXTRA_DRV_DIR}")
 message("-- Extra driver tar exclude rules: ${XDNA_EXTRA_DRV_EXCLUDES}")
+
+if(NOT SKIP_KMOD)
 add_custom_command(
   OUTPUT all_driver_source
   COMMENT "Collect all driver source code"
@@ -96,16 +98,23 @@ install(FILES ${CMAKE_CURRENT_BINARY_DIR}/dkms.conf
   COMPONENT ${XDNA_COMPONENT}
   )
 
+# install .ko for testing
+install(FILES ${XDNA_DRV_PATH} DESTINATION ${XDNA_BIN_DIR}/driver)
+
 set(amdxdna_drv_tools
   ${CMAKE_CURRENT_SOURCE_DIR}/tools/dkms_driver.sh
   ${CMAKE_CURRENT_SOURCE_DIR}/tools/npu_perf_trace.sh
   ${CMAKE_CURRENT_SOURCE_DIR}/tools/npu_perf_analyze.sh
   )
+else()
+set(amdxdna_drv_tools
+  ${CMAKE_CURRENT_SOURCE_DIR}/tools/npu_perf_trace.sh
+  ${CMAKE_CURRENT_SOURCE_DIR}/tools/npu_perf_analyze.sh
+  )
+endif()
 install(FILES ${amdxdna_drv_tools}
   PERMISSIONS OWNER_READ OWNER_WRITE OWNER_EXECUTE GROUP_READ GROUP_EXECUTE WORLD_READ WORLD_EXECUTE
   DESTINATION xrt/${XDNA_COMPONENT}
   COMPONENT ${XDNA_COMPONENT}
   )
 
-# install .ko for testing
-install(FILES ${XDNA_DRV_PATH} DESTINATION ${XDNA_BIN_DIR}/driver)
-- 
2.49.0

